source/fe/mapping_q_generic.cc

```diff
@@ -3860,62 +3883,83 @@ MappingQGeneric<dim,spacedim>::
-    a[i] = cell->vertex(i);
+    a.push_back(cell->vertex(i));
 
-  if (this->polynomial_degree>1)
-    switch (dim)
-      {
-      case 1:
-        add_line_support_points(cell, a);
-        break;
-      case 2:
-        // in 2d, add the points on the four bounding lines to the exterior
-        // (outer) points
-        add_line_support_points(cell, a);
-
-        // then get the support points on the quad if we are on a
-        // manifold, otherwise compute them from the points around it
-        if (dim != spacedim)
-          add_quad_support_points(cell, a);
-        // TODO: use get_intermediate_points_on_object for the else case
-        // unconditionally as soon as the interface of Boundary is fixed
-        else if (dynamic_cast<const Boundary<dim,spacedim> *>(&cell->get_manifold()) == NULL)
-          {
-            std::vector<Point<spacedim> > quad_points (Utilities::fixed_power<dim>(polynomial_degree-1));
-            get_intermediate_points_on_object(cell->get_manifold(), line_support_points, cell, quad_points);
-            for (unsigned int i=0; i<quad_points.size(); ++i)
-              a.push_back(quad_points[i]);
-          }
-        else
-          add_weighted_interior_points (support_point_weights_on_quad, a);
-        break;
-
-      case 3:
-      {
-        // in 3d also add the points located on the boundary faces
-        add_line_support_points (cell, a);
-        add_quad_support_points (cell, a);
+  if (this->polynomial_degree > 1)
+    {
+      // check if all entities have the same manifold id which is when we can
+      // simply ask the manifold for all points
+      Assert(dim<=3, ExcImpossibleInDim(dim));
+      bool all_manifold_ids_are_equal = (dim == spacedim);
+      if (dim > 1)
+        for (unsigned int l=0; l<GeometryInfo<dim>::lines_per_cell; ++l)
+          if (&cell->line(l)->get_manifold() != &cell->get_manifold())
+            all_manifold_ids_are_equal = false;
+      if (dim == 3)
+        if (check_identical_manifolds_of_quads(cell) == false)
+          all_manifold_ids_are_equal = false;
+      if (all_manifold_ids_are_equal)
+        {
+          std::vector<Point<spacedim> > vertices(a);
+          cell->get_manifold().add_new_points(vertices, support_point_weights_cell, a);
+          return a;
+        }
 
-        // then compute the interior points
-        // TODO: remove else case as soon as boundary is fixed
-        if (dynamic_cast<const Boundary<dim,spacedim> *>(&cell->get_manifold()) == NULL)
-          {
-            std::vector<Point<spacedim> > hex_points (Utilities::fixed_power<dim>(polynomial_degree-1));
-            get_intermediate_points_on_object(cell->get_manifold(), line_support_points, cell, hex_points);
-            for (unsigned int i=0; i<hex_points.size(); ++i)
-              a.push_back(hex_points[i]);
-          }
-        else
-          add_weighted_interior_points (support_point_weights_on_hex, a);
-        break;
-      }
+      switch (dim)
```

Comment:
this is not wrong but hard to read -- I was expecting an `else` in front of this line. It is unnecessary because you're `return`ing from the previous block, but for readability it may be nicer to enclose this block in an `else` branch anyway.